// prepack/lib/values/BooleanValue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _index = require("./index.js");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright (c) 2017-present, Facebook, Inc.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * This source code is licensed under the BSD-style license found in the
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * LICENSE file in the root directory of this source tree. An additional grant
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * of patent rights can be found in the PATENTS file in the same directory.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var BooleanValue = function (_PrimitiveValue) {
  _inherits(BooleanValue, _PrimitiveValue);

  function BooleanValue(realm, value, intrinsicName) {
    var _ret, _ret2;

    _classCallCheck(this, BooleanValue);

    var _this = _possibleConstructorReturn(this, (BooleanValue.__proto__ || Object.getPrototypeOf(BooleanValue)).call(this, realm, intrinsicName));

    _this.value = value;

    if (value && realm.intrinsics.true) return _ret = realm.intrinsics.true, _possibleConstructorReturn(_this, _ret);
    if (!value && realm.intrinsics.false) return _ret2 = realm.intrinsics.false, _possibleConstructorReturn(_this, _ret2);
    return _this;
  }

  _createClass(BooleanValue, [{
    key: "equals",
    value: function equals(x) {
      return x instanceof BooleanValue && this.value === x.value;
    }
  }, {
    key: "getHash",
    value: function getHash() {
      return this.value ? 12484058682847432 : 3777063795205331;
    }
  }, {
    key: "mightBeFalse",
    value: function mightBeFalse() {
      return !this.value;
    }
  }, {
    key: "throwIfNotConcreteBoolean",
    value: function throwIfNotConcreteBoolean() {
      return this;
    }
  }, {
    key: "_serialize",
    value: function _serialize() {
      return this.value;
    }
  }, {
    key: "toDisplayString",
    value: function toDisplayString() {
      return this.value.toString();
    }
  }]);

  return BooleanValue;
}(_index.PrimitiveValue);

exports.default = BooleanValue;
