// prepack/lib/values/SymbolValue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _index = require("./index.js");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright (c) 2017-present, Facebook, Inc.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * This source code is licensed under the BSD-style license found in the
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * LICENSE file in the root directory of this source tree. An additional grant
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * of patent rights can be found in the PATENTS file in the same directory.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var SymbolValue = function (_PrimitiveValue) {
  _inherits(SymbolValue, _PrimitiveValue);

  function SymbolValue(realm, desc, intrinsicName) {
    _classCallCheck(this, SymbolValue);

    var _this = _possibleConstructorReturn(this, (SymbolValue.__proto__ || Object.getPrototypeOf(SymbolValue)).call(this, realm, intrinsicName));

    _this.$Description = desc;
    return _this;
  }

  _createClass(SymbolValue, [{
    key: "equals",
    value: function equals(x) {
      return x instanceof SymbolValue && this.hashValue === x.hashValue && this.$Description === x.$Description;
    }
  }, {
    key: "getHash",
    value: function getHash() {
      if (!this.hashValue) {
        this.hashValue = this.$Description ? this.$Description.getHash() : ++this.$Realm.symbolCount;
      }
      return this.hashValue;
    }
  }, {
    key: "mightBeFalse",
    value: function mightBeFalse() {
      return false;
    }
  }, {
    key: "throwIfNotConcreteSymbol",
    value: function throwIfNotConcreteSymbol() {
      return this;
    }
  }, {
    key: "_serialize",
    value: function _serialize() {
      return Symbol(this.$Description);
    }
  }, {
    key: "toDisplayString",
    value: function toDisplayString() {
      if (this.$Description) {
        if (this.$Description instanceof _index.PrimitiveValue) {
          return "Symbol(" + this.$Description.toDisplayString() + ")";
        }
      }
      return "Symbol(to be supported)";
    }
  }]);

  return SymbolValue;
}(_index.PrimitiveValue);

exports.default = SymbolValue;
