// prepack/lib/singletons.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setEnvironment = setEnvironment;
exports.setFunctions = setFunctions;
exports.setJoin = setJoin;
exports.setPath = setPath;
exports.setProperties = setProperties;
var Environment = exports.Environment = null; /**
                                               * Copyright (c) 2017-present, Facebook, Inc.
                                               * All rights reserved.
                                               *
                                               * This source code is licensed under the BSD-style license found in the
                                               * LICENSE file in the root directory of this source tree. An additional grant
                                               * of patent rights can be found in the PATENTS file in the same directory.
                                               */

var Functions = exports.Functions = null;
var Join = exports.Join = null;
var Path = exports.Path = null;
var Properties = exports.Properties = null;

function setEnvironment(singleton) {
  exports.Environment = Environment = singleton;
}

function setFunctions(singleton) {
  exports.Functions = Functions = singleton;
}

function setJoin(singleton) {
  exports.Join = Join = singleton;
}

function setPath(singleton) {
  exports.Path = Path = singleton;
}

function setProperties(singleton) {
  exports.Properties = Properties = singleton;
}
