// prepack/lib/debugger/Breakpoint.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

var Breakpoint = exports.Breakpoint = function Breakpoint(filePath, line) {
  var column = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  var temporary = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
  var enabled = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : true;

  _classCallCheck(this, Breakpoint);

  this.filePath = filePath;
  this.line = line;
  this.temporary = temporary;
  this.enabled = enabled;
  this.column = column;
}

//real breakpoint set by client or temporary one set by debugger
;
